### PR TITLE
Expand private members if necessary

### PR DESCRIPTION
**Pull request by odersky**
Private members that are accessed from out of class, or that are overridden by implementation classes get mangled names and their Private flag is reset. NotJavaPrivate is eliminated.

Review by @DarkDimius

### CODE REVIEW

**Comment by xeno-by** [comment]
Would we be able to use the same mechanism to allow macro expansions to access private members visible at the macro definition site?

File: src/dotty/tools/dotc/transform/ExpandPrivate.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import Decorators._
+import ast.Trees._
+import TreeTransforms._
+
+/** Makes private methods static, provided they not deferred, accessors, or static,
+ *  by rewriting a method `m` in class `C` as follows:
+ *
+ *     private def m(ps) = e
+ *
+ *     --> private static def($this: C, ps) = [this -> $this] e
```
Comment (smarter) [inline src/dotty/tools/dotc/transform/ExpandPrivate.scala]: This description was copy-pasted from `PrivateToStatic` and does not match what this phase does.

**Comment by odersky** [comment]
You mean, make give private members accessed by a non-private macro expanded names? Yes that could work.

**Comment by DarkDimius** [comment]
@xeno-by I guess it would work only if macro expansion happens in same compilation unit where private member is defined.

**Comment by DarkDimius** [comment]
> or that are overridden by implementation classes

How does this play with separate compilation? If you have a implementation class, that overrides a private field but is separately compiled, will you try to retroactively mangle field defined in super class?

**Comment by DarkDimius** [comment]
By following the code I do not see how it will trigger problems.

**Comment by DarkDimius** [comment]
@odersky, there are other private methods that need to be `expanded`: private deferred methods in traits, such as super accessors. Please review last commit.

**Comment by DarkDimius** [comment]
Last fix allows dotty to pass flag verification of JVM.

**Comment by DarkDimius** [comment]
Ok, I'm preparing a fix.

**Comment by odersky** [comment]
It seems there's more wrong with superaccessors than this. You need to fix
it there. Patches later on won't help. I am fairly close to a solution, but
the fix uncovers further errors.


On Sun, May 3, 2015 at 7:58 PM, Dmitry Petrashko <notifications@github.com>
wrote:

> Ok, I'm preparing a fix.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/pull/517#issuecomment-98517671>.
>



-- 
Martin Odersky
EPFL

**Comment by DarkDimius** [comment]
@odersky please review last 3 commits.

**Comment by odersky** [comment]
LGTM. I had the same as your first commit, but got errors that were solved by your third commit. I am still puzzled why the scalac SuperAccessors explicitly defines symbols that are both Deferred and Private. I am also puzzled where super accessors for non traits (which are now private, but not deferred) actually get defined. But we'll figure that out, eventually.

**Comment by odersky** [comment]
I figured it out. resolveSuper does the right thing for private super accessors. I'll add a test.

**Comment by DarkDimius** [comment]
@odersky, I have non-commited post-condition for super-accessors that checks that private super-accessors have non-empty rhs. I'll make a PR.

**Pull request merged**